Brightspace/insights-engagement-dashboard
# Us117429/temp simple role filter

Quality notes
* Testing
  * Unit tests for roles model class and insights-role-filter component
  * Manual testing
    * It loads the roles from the LMS using the `GET /d2l/api/lp/(version)/roles/` endpoint
    * It orders the roles by role name (and role id if multiple roles have the same name)
    * Every time a role is selected or unselected in the filter, it fires an event that the dashboard can listen for
* Security
  * Should only load roles that the dashboard user can manage (based on the notes in the valence docs)
* Devops
  * No error handling. 404 causes the filter to load with no entries inside
  * It uses a temporary `simple-filter` component that will eventually be replaced with [`facet-filter-sort`](https://github.com/BrightspaceUI/facet-filter-sort) once the latest version of that component is available in BSI
* Perf: 
  * Depends on valence API perf
  * Sorting roles by name may be an expensive operation. I haven't tested with lots of roles
* UX/Docs
  * Added a11y unit test, did not test a11y manually
  * No localization support yet

Screenshot
![image](https://user-images.githubusercontent.com/11587338/88721454-79115c80-d0f4-11ea-8415-fe2bfcac7171.png)

FYI @hyehayes @rohitvinnakota @anhill-D2L @MykolaGalian 

--------------------------

*Pre-review discussion*
Looking to have a discussion around these changes before I submit this for actual review. Here are the questions I'm hoping to answer:
* I made `rolesData` a property of `engagement-dashboard` so that whenever it was updated, it would re-render the roles filter. 
  * Is that the right approach? 
  * Or should I convert roles to a mobxLitElement so that triggering a re-render is handled by mobx? 
  * Or is there some other way of triggering a re-render that I'm not aware of?
* `roles.js` is similar to our I/O components in SSR, so it's tricky to test.
  * Does this component need testing? (I think so, looking for other opinions)
  * If it does, how should I go about that?
    * My thoughts: I think mocking the `fetch` API is not a great idea. I was thinking the karma test server could probably serve up test data, but I'm not sure how to configure it to do that.
